integration tests

DieracDelta opened the pull request:
This PR addresses #237, and adds in a couple of changes:

- The use of functions to check if test output is correct for plugin tests. These have been added for:
  - nqueens
  - 2 by 2 sudoku
  - hash function
- Colored output. This makes the test output a bit easier for me to read. Do yall think this is useful? If not, I can easily revert.
- The plugin tests have been split out into unit tests and integration tests in the same style as the unit and performance tests in the wp library. Additional makefile targets have been added; integration tests are not run by default (`make test`) as they take a while and include a couple timeouts.
    - To run these integration tests, `make test.plugin.integration` from the `wp` directory or `make test.integration` from the `plugin` directory.
- Two test changes have been added:
  - Multiple nqueens tests ranging from 4 to 19 inclusive. This adds in a
    dependency on clang. However, this should be easy to install (`apt-get
    install clang`). Do we have a place to list dependencies?
  - A 3 by 3 sudoku solver implementation. Unfortunately, the constraint encoding is too slow (and will time out). This is not surprising, as it encodes many more constraints than n-queens. I've got it as skipped right now, but maybe it makes sense to move this to the private repo instead.

fortunac commented on wp/lib/bap_wp/Makefile:
```diff
@@ -68,11 +68,11 @@ test: test.unit test.performance
 .PHONY: test.unit
 test.unit:
-	dune runtest tests/unit
+	OUINT_CI=true dune runtest tests/unit
```
What does this env var do? Should it be `OUNIT_CI`?

fortunac commented on wp/plugin/tests/test_libs/test_wp.ml:
```diff
@@ -0,0 +1,577 @@
+       ~checker:(check_n_queens i `x86_64 |> Some)
+       ~script:script ~args:[string_of_int i]))
+
+let integration_tests = List.append [
```
This file is getting rather large. What are your thoughts on separating the unit tests and integration tests into two different files?

fortunac commented on wp/plugin/tests/test_libs/test_wp.ml:
```diff
@@ -0,0 +1,577 @@
+open OUnit2
+open OUnitTest
+
+module StringMap = Map.Make(String);;
```
This is a little nit-pick, but why do we need the `;;` over here?

codyroux replied:
Don't use `;;` in `.ml` files, as a rule.

DieracDelta replied:
Good point! Seems to compile fine without `;;`, so I'll remove them.

fortunac commented:
Tests look good! I'm currently getting three timeouts on the integration tests, on nqueens 17, 18, and 19. Is this expected? If not, what are your thoughts an allowing the use to specify their own timeout when invoking the tests? We can leave that for a separate PR since it's a separate issue.

DieracDelta commented:
Based on @fortunac's feedback, I've made some changes:

- Removed unnecessary semicolons
- Skipped n=17-19 in nqueens, since these test cases time out right now.
- Split `test_wp.ml` into several files:
  - `test_wp_utils.ml` which contains helper functions used by both the unit and integration tests.
  - `test_wp_unit.ml` which declares the unit test suite.
  - `test_wp_integration.ml` which declares the integration test suite.

fortunac commented on docs/installation.html:
```diff
@@ -72,12 +72,24 @@ <h3>Installing CBAT</h3>
   make test
 </pre>
 
+<p>To run oUnit integration tests:
```
Can we add this to the site?

fortunac reviewed (commented):
Looks good! After ~~adding that target to the site and~~ rebasing, I have no further changes requested. 🎉
